File: src/day03.py

def build_candidates(data,deltas):
    (dx,dy) = deltas
    rows = len(data)
    c = [ (r*dy,r*dx) for  r in range(rows) if r*dy < rows]
    return c

def test_candidate(data,loc):
    (row,col)=loc
    num_rows = len(data)
    num_cols = len(data[0])-1
    new_col = col % num_cols
    row_string = data[row]
    c = row_string[new_col]
    if c=='#':
        return 1
    else:
        return 0


data = open("../data/day03.dat").readlines()

candidates = build_candidates(data,[3,1])

def part1(data,candidates):
    return sum([test_candidate(data, x) for x in candidates ])
slopes= [ [1, 1], [3, 1], [5, 1], [7, 1], [1, 2]]

x= [build_candidates(data, deltas) for deltas in slopes]
y = [part1(data, c) for c in x]

File: src/day03.clj
(ns day03
  (:require [clojure.java.io :as io]
            [clojure.string :as str]
            [hashp.core]
            ))

(def data  (->> (io/resource "day03.dat")
                slurp
                str/split-lines))

(defn build-candidates [data [dx dy]]
  (let [rows (count data)]
    (for [r (range rows)
          :when (<= (* r dy) rows)
          ]
      [(* r dy) (* r dx) ])))

(defn test-candidate [data [row col] ]
  (let [num-rows (count data)
        num-cols (count (first data))
        new-col (mod col num-cols)
        row_string (nth data row)
        c (nth row_string new-col)
        ]
    c
    ))


(defn part1 [data slope]
  (->> (build-candidates data slope)
       (map (partial test-candidate data))
       (filter #(= % \#))
       count
       ))

(part1 data [3 1])
(def slopes [ [1 1] [3 1] [5 1] [7 1] [1 2]])

(defn part2 [data slopes]
(->>
 (map (partial part1 data) slopes)
 (reduce *)
 ))

(part2 data slopes)
